[Docs] Replace jupyter with interactive dev (#3516)
skypilot-org/skypilot: docs/source/examples/interactive-development.rst

Start a Development Cluster
===========================

SkyPilot makes the interactive development very easy on clouds. It helps you:

#. Launch: a cluster with GPU or any specific requirement with a single command.

> #. Connect: Quickly connect to the cluster with the cluster name.

Reviewer: This ordering is different from sections below?

Author: Moved autostop earlier as it might be more like a value add : )

> #. Launch: a cluster with GPU or any specific requirement with a single command.
> #. Connect: Quickly connect to the cluster with the cluster name.
> #. Autostop: Automatically stop the cluster after a period of time for cost savings.

Reviewer: ```suggestion
#. Launch: Quickly get a cluster with GPU or other resource requirement with a single command.
#. Connect: Easily connect to the cluster using the cluster name (e.g., ``ssh dev``).
#. Autostop: Automatically stop the cluster after some idle time for cost savings.
```

Launch
------

To launch a cluster with a cheap GPU for developemnt:

.. code-block:: bash

>   # Launch a cluster with 1 NVIDIA GPU and syncing the local working directory to the
>   # cluster.

Reviewer: ```suggestion
  # Launch a cluster with 1 NVIDIA GPU and sync the local working directory to the
  # cluster. This can be launched as a pod in your Kubernetes cluster or a cloud VM.
```

>   sky launch -c dev --gpus T4 --workdir .

Reviewer: Idea: to make "This can be launched as a pod in your Kubernetes cluster or a cloud VM." more clear, we can take it out of comments, and put it as a normal text, followed by tabbed screenshots: K8s -> a screenshot of a pod named `sky .. dev` in K8s outputs/dashboard; cloud -> a screenshot of cloud console, etc.

Wdyt?

Author: Added. Thanks!

.. note::

  View the supported GPUs with the :code:`sky show-gpus` command.

Autostop
--------

> SkyPilot allows you to automatically stop the cluster after a period of time to save costs. You can set the autostop time with a single command.

Reviewer: ```suggestion
SkyPilot allows you to automatically stop the cluster after a period of idle time to save costs. You can set the autostop time with a single command:
```

.. code-block:: bash

  # Auto stop the cluster after 5 hours
  sky autostop -i 300 dev

Or add an additional flag :code:`-i` during the launch:

.. code-block:: bash

  # Launch a cluster with auto stop after 5 hours
  sky launch -c dev --gpus T4 --workdir . -i 300

> For more details of auto stopping, check out: :ref:`auto-stop`. It helps you avoid idle machines from costing you a fortune, when you want a
> machine to stop automatically after you go to bed or forget to stop it.

Reviewer: ```suggestion
For more details, see :ref:`auto-stop`.
```

I found the second sentence a bit distracting (not sure about its core message?).

Author: Was trying to explicitly give some concrete examples for why people want the autostopping, as we have heard a bunch of people talking about letting the VM up for a while after they go to bed.

Connect
-------

> A user can easily connect to the cluster with the following ways:

Reviewer: ```suggestion
A user can easily connect to the cluster using your method of choice:
```

SSH
~~~

> SkyPilot will automatically configure the SSH setting for a user to connect to the cluster with the cluster name after it is launched.

Reviewer: ```suggestion
SkyPilot will automatically configure the SSH setting for a cluster, so that users can connect to the cluster with the cluster name:
```

.. code-block:: bash
  
  ssh dev

VSCode
~~~~~~

> A common usecase for interactive developement is to connect an IDE directly to the cluster and edit code directly on the cluster. It
> can also simply done by connecting VSCode to the cluster with the cluster name:

Reviewer: ```suggestion
A common use case for interactive development is to connect a local IDE to a remote cluster and directly edit code that lives on the cluster. 
This is supported by simply connecting VSCode to the cluster with the cluster name:
```

#. Click on the top bar, type: :code:`> remote-ssh`, and select :code:`Remote-SSH: Connect Current Window to Host...`

> #. Select the cluster name from the list of hosts.

Reviewer: ```suggestion
#. Select the cluster name (e.g., ``dev``) from the list of hosts.
```

#. Open folder: :code:`sky_workdir` and find your code.

.. image:: https://imgur.com/8mKfsET.gif
  :align: center
  :alt: Connect to the cluster with VSCode

Jupyter Notebooks
~~~~~~~~~~~~~~~~~

Jupyter notebooks are a useful tool for interactive development, debugging, and
visualization.

To get a machine with a GPU attached, use:

.. code-block:: bash

>    ssh -L 8888:localhost:8888 dev

Reviewer: fix?

Author: Fixed the wording above : )

  sky autostop -i 300 dev

Or add an additional flag :code:`-i` during the launch:

.. code-block:: bash

  # Launch a cluster with auto stop after 5 hours
  sky launch -c dev --gpus T4 --workdir . -i 300

For more details of auto stopping, check out: :ref:`auto-stop`. This feature is designed
to prevent idle clusters from incurring unnecessary costs, ensuring your cluster
stops automatically, whether it's overnight or throughout the weekend.

.. _dev-connect:

Connect
-------

A user can easily connect to the cluster using your method of choice:

.. _dev-ssh:

SSH
~~~

SkyPilot will automatically configure the SSH setting for a cluster, so that users can connect to the cluster with the cluster name:

.. code-block:: bash
  
  ssh dev

.. _dev-vscode:

VSCode
~~~~~~

A common use case for interactive development is to connect a local IDE to a remote cluster and directly edit code that lives on the cluster. 
This is supported by simply connecting VSCode to the cluster with the cluster name:

#. Click on the top bar, type: :code:`> remote-ssh`, and select :code:`Remote-SSH: Connect Current Window to Host...`
#. Select the cluster name (e.g., ``dev``) from the list of hosts.

> #. Open folder: :code:`sky_workdir` and find your code.

Reviewer 2: Minor suggestion, feel free to ignore - this one-liner `code --remote ssh-remote+<cluster_name> '/home/sky/sky_workdir'` may be easier to use than clicking in menus

Author: Good point! It seems that we can not easily determine the home path for '/home/sky/sky_workdir' on different clouds. I am going to leave this out for now. Will try to add this later.
